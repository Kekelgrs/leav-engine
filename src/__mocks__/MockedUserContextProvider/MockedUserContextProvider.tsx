import React, {ReactElement} from 'react';
import UserContext, {IUserContext} from '../../components/shared/UserContext/UserContext';
import {PermissionsActions} from '../../_gqlTypes/globalTypes';

interface IProps {
    permissions?: {[permName: string]: boolean};
    children: ReactElement;
}
function MockedUserContextProvider({permissions, children}: IProps) {
    const userData: IUserContext = {
        id: 1,
        name: 'Test',
        permissions:
            permissions ??
            Object.values(PermissionsActions)
                .filter(a => !!a.match(/^app_/))
                .reduce((perms, p) => {
                    perms[p] = true;

                    return perms;
                }, {})
    };

    return <UserContext.Provider value={userData}>{children}</UserContext.Provider>;
}

export default MockedUserContextProvider;
