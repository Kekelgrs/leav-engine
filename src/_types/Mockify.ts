// Used to mock any interface, turning all function properties to an optionnal mock
// Mockified object must be then passed to a function with a type assertion
export type Mockify<T> = {[P in keyof T]?: T[P] extends (...args: any) => any ? jest.Mock : T[P]};
