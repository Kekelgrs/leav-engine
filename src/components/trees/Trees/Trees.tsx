import {History} from 'history';
import React, {useState} from 'react';
import {useTranslation} from 'react-i18next';
import {Link} from 'react-router-dom';
import {Button, Grid, Header, Icon} from 'semantic-ui-react';
import useLang from '../../../hooks/useLang';
import useUserData from '../../../hooks/useUserData';
import {getTreesQuery, TreesQuery} from '../../../queries/trees/getTreesQuery';
import {addWildcardToFilters} from '../../../utils/utils';
import {GET_TREESVariables} from '../../../_gqlTypes/GET_TREES';
import {PermissionsActions} from '../../../_gqlTypes/globalTypes';
import TreesList from '../TreesList';

interface ITreesProps {
    history: History;
}

/* tslint:disable-next-line:variable-name */
const Trees = ({history}: ITreesProps): JSX.Element => {
    const {t} = useTranslation();
    const {lang} = useLang();
    const [filters, setFilters] = useState<Partial<GET_TREESVariables>>({});
    const userData = useUserData();

    const _onFiltersUpdate = (filterElem: any) => {
        const newElemState =
            filterElem.type === 'checkbox'
                ? filterElem.indeterminate
                    ? undefined
                    : filterElem.checked
                : filterElem.value;

        setFilters({
            ...filters,
            [filterElem.name]: newElemState
        });
    };

    return (
        <>
            <Grid>
                <Grid.Column textAlign="left" floated="left" width={8} verticalAlign="middle">
                    <Header size="large">
                        <Icon name="share alternate" />
                        {t('trees.title')}
                    </Header>
                </Grid.Column>
                {userData.permissions[PermissionsActions.admin_create_tree] && (
                    <Grid.Column floated="right" width={3} textAlign="right" verticalAlign="middle">
                        <Button icon labelPosition="left" size="medium" as={Link} to={'/trees/edit/'}>
                            <Icon name="plus" />
                            {t('trees.new')}
                        </Button>
                    </Grid.Column>
                )}
            </Grid>
            <TreesQuery query={getTreesQuery} variables={{...addWildcardToFilters(filters), lang}}>
                {({loading, error, data}) => {
                    if (typeof error !== 'undefined') {
                        return <p>Error : {error.message}</p>;
                    }

                    const onRowClick = tree => history.push('/trees/edit/' + tree.id);

                    return (
                        <TreesList
                            trees={data && data.trees ? data.trees.list : null}
                            onRowClick={onRowClick}
                            onFiltersUpdate={_onFiltersUpdate}
                            filters={filters}
                            loading={loading}
                        />
                    );
                }}
            </TreesQuery>
        </>
    );
};

export default Trees;
