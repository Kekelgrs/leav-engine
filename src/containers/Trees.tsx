import * as React from 'react';
import {Header, Icon} from 'semantic-ui-react';

function Trees() {
    return (
        <Header size="large">
            <Icon name="share alternate" />
            Trees
        </Header>
    );
}

export default Trees;
