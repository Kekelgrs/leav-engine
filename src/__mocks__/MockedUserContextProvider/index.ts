import MockedUserContextProvider from './MockedUserContextProvider';
export default MockedUserContextProvider;
