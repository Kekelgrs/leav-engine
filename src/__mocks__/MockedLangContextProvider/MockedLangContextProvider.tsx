import React from 'react';
import LangContext from '../../components/shared/LangContext';
import {ILangContext} from '../../components/shared/LangContext/LangContext';
import {AvailableLanguage} from '../../_types/types';

function MockedLangContextProvider({children}: any) {
    const mockLangs: ILangContext = {
        lang: [AvailableLanguage.fr],
        availableLangs: [AvailableLanguage.fr, AvailableLanguage.en],
        defaultLang: AvailableLanguage.fr
    };

    return <LangContext.Provider value={mockLangs}>{children}</LangContext.Provider>;
}

export default MockedLangContextProvider;
