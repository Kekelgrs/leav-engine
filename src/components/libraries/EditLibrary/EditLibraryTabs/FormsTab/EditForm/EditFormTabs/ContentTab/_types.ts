import {DragObjectWithType} from 'react-dnd';
import {GET_ATTRIBUTES_attributes_list} from '../../../../../../../../_gqlTypes/GET_ATTRIBUTES';
import {FormElementTypes} from '../../../../../../../../_gqlTypes/globalTypes';
import {IKeyValue} from '../../../../../../../../_types/shared';

export const PLACEHOLDER_ID = '__placeholder__';

export enum UIElementTypes {
    DIVIDER = 'divider',
    FIELDS_CONTAINER = 'fields_container',
    TEXT_BLOCK = 'text_block',
    TABS = 'tabs'
}

export enum FieldTypes {
    TEXT_INPUT = 'input_field',
    DATE = 'date',
    CHECKBOX = 'checkbox',
    ENCRYPTED = 'encrypted',
    DROPDOWN = 'dropdown'
}

export enum DraggableElementTypes {
    RESERVE_LAYOUT_ELEMENT = 'RESERVE_LAYOUT_ELEMENT',
    ATTRIBUTE = 'ATTRIBUTE',
    FORM_ELEMENT = 'FORM_ELEMENT'
}

export interface IFormElement {
    id: string;
    order: number;
    type: FormElementTypes;
    uiElement: IUIElement;
    containerId: string;
    settings?: IKeyValue<any>;
    herited?: boolean;
}

export interface IUIElement {
    type: UIElementTypes | FieldTypes;
    component: JSX.Element;
    canDrop: (dropCandidate: IFormElement) => boolean;
    settings?: string[];
}

export interface IFormElementProps<T extends object> {
    elementData?: IFormElement;
    settings: T;
}

export interface ICommonFieldsSettings {
    label?: string;
    attribute?: GET_ATTRIBUTES_attributes_list;
}

export interface IFormElementPos {
    order: number;
    containerId: string;
}

export interface IFormBuilderDragObject<T extends IUIElement | IFormElement> extends DragObjectWithType {
    type: DraggableElementTypes;
    element: T;
    index: number;
    currentPos?: IFormElementPos;
    originPos?: IFormElementPos;
    dropAtPos?: IFormElementPos;
}
