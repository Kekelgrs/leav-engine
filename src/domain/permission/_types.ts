import {
    AppPermissionsActions,
    IPermissionsTreeTarget,
    LibraryPermissionsActions,
    PermissionsActions,
    PermissionTypes
} from '_types/permissions';
import {IQueryInfos} from '_types/queryInfos';
import {ITreeNode} from '_types/tree';

export interface IPermissionTarget {
    attributeId?: string;
    recordId?: string;
}

export interface IGetSimplePermissionsParams {
    type: PermissionTypes;
    applyTo: string | null;
    action: PermissionsActions;
    usersGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IGetPermissionsByActionsParams {
    type: PermissionTypes;
    applyTo: string | null;
    actions: PermissionsActions[];
    usersGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export type PermByActionsRes = {[name: string]: boolean | null} | null;

export interface IGetPermissionByUserGroupsParams {
    type: PermissionTypes;
    action: PermissionsActions;
    userGroupsPaths: ITreeNode[][];
    applyTo?: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IGetAppPermissionParams {
    action: AppPermissionsActions;
    userId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedAppPermissionParams {
    action: AppPermissionsActions;
    userGroupId: string;
    ctx: IQueryInfos;
}

export interface IGetLibraryPermissionParams {
    action: LibraryPermissionsActions;
    libraryId: string;
    userId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedLibraryPermissionParams {
    action: LibraryPermissionsActions;
    libraryId: string;
    userGroupId: string;
    ctx: IQueryInfos;
}

export interface IGetHeritedPermissionsParams {
    type: PermissionTypes;
    applyTo: string;
    action: PermissionsActions;
    userGroupId: string;
    permissionTreeTarget?: IPermissionsTreeTarget;
    ctx: IQueryInfos;
}

export interface IIsAllowedParams {
    type: PermissionTypes;
    action: PermissionsActions;
    userId: string;
    applyTo?: string;
    target?: IPermissionTarget;
    ctx: IQueryInfos;
}
