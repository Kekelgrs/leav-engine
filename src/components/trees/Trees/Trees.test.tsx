// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {MockedProvider} from '@apollo/react-testing';
import {render} from 'enzyme';
import {History} from 'history';
import React from 'react';
import {BrowserRouter as Router} from 'react-router-dom';
import {Mockify} from '../../../_types//Mockify';
import MockedLangContextProvider from '../../../__mocks__/MockedLangContextProvider';
import MockedUserContextProvider from '../../../__mocks__/MockedUserContextProvider';
import Trees from './Trees';

describe('Trees', () => {
    test('Snapshot test', async () => {
        const mockHistory: Mockify<History> = {};

        const comp = render(
            <MockedProvider>
                <MockedLangContextProvider>
                    <MockedUserContextProvider>
                        <Router>
                            <Trees history={mockHistory as History} />
                        </Router>
                    </MockedUserContextProvider>
                </MockedLangContextProvider>
            </MockedProvider>
        );

        expect(comp).toMatchSnapshot();
    });
});
