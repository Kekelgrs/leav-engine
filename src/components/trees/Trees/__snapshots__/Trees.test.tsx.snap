// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trees Snapshot test 1`] = `
<div
  className="ui large header"
>
  <i
    aria-hidden="true"
    className="share alternate icon"
  />
  trees.title
</div>
`;
